src/desktop/state/FocusState.cpp

```diff
@@ -0,0 +1,322 @@
+#include "../../xwayland/XSurface.hpp"
+#include "../../protocols/PointerConstraints.hpp"
+
+using namespace Desktop;
+
+SP<CFocusState> Desktop::focusState() {
+    static SP<CFocusState> state = makeShared<CFocusState>();
+    return state;
+}
+
+Desktop::CFocusState::CFocusState() {
+    m_windowOpen = g_pHookSystem->hookDynamic("openWindowEarly", [this](void* self, SCallbackInfo& info, std::any data) {
+        auto window = std::any_cast<PHLWINDOW>(data);
+
+        addWindowToHistory(window);
+    });
+
+    m_windowClose = g_pHookSystem->hookDynamic("closeWindow", [this](void* self, SCallbackInfo& info, std::any data) {
+        auto window = std::any_cast<PHLWINDOW>(data);
+
+        removeWindowFromHistory(window);
+    });
+}
+
+static void onFullscreenWorkspaceFocusWindow(PHLWINDOW pWindow, bool forceFSCycle) {
+    const auto FSWINDOW = pWindow->m_workspace->getFullscreenWindow();
+    const auto FSMODE   = pWindow->m_workspace->m_fullscreenMode;
+
+    if (pWindow == FSWINDOW)
+        return; // no conflict
+
+    if (pWindow->m_isFloating) {
+        // if the window is floating, just bring it to the top
+        pWindow->m_createdOverFullscreen = true;
+        g_pHyprRenderer->damageWindow(pWindow);
+        return;
+    }
+
+    static auto PONFOCUSUNDERFS = CConfigValue<Hyprlang::INT>("misc:on_focus_under_fullscreen");
+
+    switch (*PONFOCUSUNDERFS) {
+        case 0:
+            // ignore, don't do anything.
+            break;
+        case 2:
+            // undo fs, unless we force a cycle
+            if (!forceFSCycle) {
+                g_pCompositor->setWindowFullscreenInternal(FSWINDOW, FSMODE_NONE);
+                break;
+            }
+            [[fallthrough]];
+        case 1:
+            // replace fullscreen
+            g_pCompositor->setWindowFullscreenInternal(FSWINDOW, FSMODE_NONE);
+            g_pCompositor->setWindowFullscreenInternal(pWindow, FSMODE);
+            break;
+
+        default: Debug::log(ERR, "Invalid misc:on_focus_under_fullscreen mode: {}", *PONFOCUSUNDERFS); break;
+    }
+}
+
+void CFocusState::fullWindowFocus(PHLWINDOW pWindow, SP<CWLSurfaceResource> surface, bool preserveFocusHistory, bool forceFSCycle) {
+    if (pWindow) {
+        if (!pWindow->m_workspace)
+            return;
+
+        const auto CURRENT_FS_MODE = pWindow->m_workspace->m_hasFullscreenWindow ? pWindow->m_workspace->m_fullscreenMode : FSMODE_NONE;
+        if (CURRENT_FS_MODE != FSMODE_NONE)
+            onFullscreenWorkspaceFocusWindow(pWindow, forceFSCycle);
+    }
+
+    static auto PMODALPARENTBLOCKING = CConfigValue<Hyprlang::INT>("general:modal_parent_blocking");
+
+    if (*PMODALPARENTBLOCKING && pWindow && pWindow->m_xdgSurface && pWindow->m_xdgSurface->m_toplevel && pWindow->m_xdgSurface->m_toplevel->anyChildModal()) {
+        Debug::log(LOG, "Refusing focus to window shadowed by modal dialog");
+        return;
+    }
+
+    rawWindowFocus(pWindow, surface, preserveFocusHistory);
+}
```

Comment:
This method does not look right: we conditionally call `onFullscreenWorkspaceFocusWindow`, but then unconditionally call `rawWindowFocus`. I suppose, `rawWindowFocus` should only be called when `onFullscreenWorkspaceFocusWindow` is not?

Still don't know why the tests pass...